docs/auto-tag-products-with-their-vendors/script.liquid

```diff
@@ -1,37 +1,106 @@
-{% if event.preview %}
-  {% action "shopify" %}
-    mutation {
-      tagsAdd(
-        id: "gid://shopify/Product/1234576890"
-        tags: "Nike"
-      ) {
-        userErrors {
-          field
-          message
+{% assign vendor_tag_prefix = options.apply_this_prefix_to_vendor_tags %}
+
+{% if event.topic == "mechanic/user/trigger" %}
+  {% assign cursor = nil %}
+  {% assign products = array %}
+
+  {% for n in (0..100) %}
+    {% capture query %}
+      query {
+        products(
+          first: 250
+          after: {{ cursor | json }}
+        ) {
+          pageInfo {
+            hasNextPage
+          }
+          edges {
+            cursor
+            node {
+              id
+              tags
+              vendor
+            }
+          }
         }
       }
-    }
-  {% endaction %}
-{% endif %}
+    {% endcapture %}
+
+    {% assign result = query | shopify %}
+
+    {% assign products_result = result.data.products.edges | map: "node" | where: "vendor" %}
+    {% assign products = products | concat: products_result %}
+
+    {% if result.data.products.pageInfo.hasNextPage %}
+      {% assign cursor = result.data.products.edges.last.cursor %}
+    {% else %}
+      {% break %}
+    {% endif %}
+  {% endfor %}
 
-{% if event.topic == "mechanic/user/trigger" %}
-  {% assign products = shop.products %}
 {% elsif event.topic contains "shopify/products/" %}
+  {% if product.vendor == blank %}
+    {% break %}
+  {% endif %}
+
+  {% assign graphql_product = hash %}
+  {% assign graphql_product["id"] = product.admin_graphql_api_id %}
+  {% assign graphql_product["tags"] = product.tags | split: ", " %}
+  {% assign graphql_product["vendor"] = product.vendor %}
+
   {% assign products = array %}
-  {% assign products[0] = product %}
+  {% assign products[0] = graphql_product %}
+{% endif %}
+
+{% if event.preview %}
+  {% capture products_json %}
+    [
+      {
+        "id": "gid://shopify/Product/1234567890",
+        "vendor": "ACME"
+      }
+    ]
+  {% endcapture %}
+
+  {% assign products = products_json | parse_json %}
 {% endif %}
 
 {% for product in products %}
-  {% if product.vendor == blank %}
-    {% continue %}
+  {% assign tags_to_remove = array %}
+  {% assign vendor_tag = product.vendor %}
+
+  {% if vendor_tag_prefix != blank %}
+    {% assign vendor_tag = vendor_tag | prepend: vendor_tag_prefix %}
+
+    {% for tag in product.tags %}
+      {% if tag contains vendor_tag_prefix and tag != vendor_tag %}
```

Comment:
`tag contains vendor_tag_prefix` will have false positives if the prefix ends up _later_ in the tag - and since we're talking about removing tags automatically here, I don't want to risk it. Can you check for the presence of the prefix at the _beginning_ of the tag string? Maybe slice the tag by the prefix length, then check for equality.

Reply:
I actually had started to write that, then was secretly hoping for a "begins_with" filter. I'll add it in

Reply:
hehehehe